File: BalancedSmileys/cpp/bs.cpp
// Balanced Smileys solution in C++ for CodeEval.com by Steven A. Dunn

#include <cmath>
#include <fstream>
#include <iostream>
#include <string>

using std::abs;
using std::cout;
using std::endl;
using std::ifstream;
using std::string;

string isBalanced(string);
string replaceSmileys(string);
void replaceAll(string&, string, string);

int main(int argc, char* argv[])
{
  ifstream inputFile(argv[1]);
  string line;

  if (inputFile)
  {
    while (getline(inputFile, line))
      cout << isBalanced(line) << endl;
    inputFile.close();
  }
  return 0;
}

string isBalanced(string line)
{
  if (line == "")
    return "YES";

  line = replaceSmileys(line);
 
  int min = 0;
  int max = 0;
  int leftParenFound = 0;
  for (int i = 0; i < line.size(); ++i)
  {
    char c = line[i];
    if (c == '(') 
    {
      ++min;
      ++max;
    }
    else if (c == ')') 
    {
      --min;
      --max;
      if (max < 0) { return "NO"; }
    }
    else if (c == '{') 
    {
      ++max;
    }
    else if (c == '}') 
    {
      --min;
    }   
  }
 
  if (min <= 0 && 0 <= max)
    return "YES"; 

  return "NO";
}

string replaceSmileys(string line)
{
  string replacedLine = line;
  replaceAll(replacedLine, ":)", "}");
  replaceAll(replacedLine, ":(", "{"); 
  return replacedLine;
}

void replaceAll(string& line, string from, string to)
{
  int start = 0;
  while((start = line.find(from, start)) != string::npos)
  {
    line.replace(start, from.length(), to);
    start + to.length();
  }
}

File: BalancedSmileys/py3/bs.py
# balanced smileys soln in py3 for code eval by steven a dunn

import sys

def is_balanced(line):
    if line == "":
        return "YES"

    line = line.replace(":)", "}").replace(":(", "{")

    min_val = 0
    max_val = 0
    for i in range(0, len(line)):
        c = line[i]
        if c == '(':
            min_val += 1
            max_val += 1
        elif c == ')':
            min_val -= 1
            max_val -= 1
            if max_val < 0:
                return "NO"
        elif c == '{':
            max_val += 1
        elif c == '}':
            min_val -= 1

    if min_val <= 0 and 0 <= max_val:
        return "YES"
    return "NO"

f = open(sys.argv[1], 'r')
for line in f:
    print (is_balanced(line))
f.close()